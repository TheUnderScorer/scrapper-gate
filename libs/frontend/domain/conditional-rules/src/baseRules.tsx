import { FrontendConditionalRuleDefinition } from './types';

export const baseRulesSelection: FrontendConditionalRuleDefinition[] = [];
