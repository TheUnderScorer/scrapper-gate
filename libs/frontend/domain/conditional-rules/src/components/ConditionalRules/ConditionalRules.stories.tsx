import { variableRule } from '../../rules/variableRule';
import { FormVariablesProvider } from '@scrapper-gate/frontend/domain/variables';
import {
  ConditionalRuleTypes,
  ConditionalRuleWhen,
} from '@scrapper-gate/shared/domain/conditional-rules';
import { createVariable } from '@scrapper-gate/shared/domain/variables';
import {
  ConditionalRuleGroupType,
  Variable,
  VariableScope,
  VariableType,
} from '@scrapper-gate/shared/schema';
import { subDays, subHours } from 'date-fns';
import React, { useMemo } from 'react';
import { Form } from 'react-final-form';
import { dateRule } from '../../rules/dateRule';
import { makeHtmlElementRule } from '../../rules/htmlRule';
import { ConditionalRules } from './ConditionalRules';

export default {
  title: 'Conditional Rules',
};

const variables: Variable[] = [
  createVariable({
    key: 'Myvariable',
    value: 'Variable test',
    defaultValue: 'Test',
    scope: VariableScope.Global,
  }),
  createVariable({
    key: 'Date',
    defaultValue: subHours(subDays(new Date(), 1), 2).toISOString(),
    scope: VariableScope.Global,
    type: VariableType.Date,
  }),
];

const conditionalRules = [
  {
    type: ConditionalRuleGroupType.All,
    rules: [
      {
        value: '{{Date}}',
        when: ConditionalRuleWhen.Equals,
        type: ConditionalRuleTypes.Date,
        id: '#id',
      },
      {
        type: ConditionalRuleTypes.HtmlElement,
      },
    ],
  },
];

export const Component = () => {
  const rules = useMemo(
    () => [
      dateRule,
      makeHtmlElementRule({
        highlightId: 'highlight',
      }),
      variableRule,
    ],
    []
  );

  return (
    <Form
      initialValues={{
        variables,
        conditionalRules,
      }}
      onSubmit={console.log}
      render={(props) => (
        <FormVariablesProvider name="variables">
          <form>
            <ConditionalRules
              fieldVariant="outlined"
              helperText="Configure rules that happen when various stuff happens."
              label="Rules"
              definitions={rules}
              name="conditionalRules"
            />
            <pre>{JSON.stringify(props.values, null, ' ')}</pre>
          </form>
        </FormVariablesProvider>
      )}
    />
  );
};
