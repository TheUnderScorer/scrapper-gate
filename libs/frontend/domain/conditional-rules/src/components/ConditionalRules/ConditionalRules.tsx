import React, { useCallback, useState } from 'react';
import {
  ConditionalRuleGroupInput,
  ConditionalRuleGroupType,
} from '@scrapper-gate/shared/schema';
import {
  FieldProps,
  useFieldArray,
  useFieldHasError,
} from '@scrapper-gate/frontend/form';
import {
  Box,
  Divider,
  Fab,
  FormHelperText,
  FormLabel,
  Stack,
  Typography,
} from '@material-ui/core';
import {
  ConditionalRulesGroup,
  ConditionalRulesGroupProps,
} from './Group/ConditionalRulesGroup';
import { Centered } from '@scrapper-gate/frontend/ui';
import { makeStyles } from '@material-ui/core/styles';
import { ConditionalRulesSelection } from '../../types';
import { Add } from '@material-ui/icons';
import classNames from 'classnames';

export interface ConditionalRulesProps
  extends FieldProps<ConditionalRuleGroupInput[]>,
    Pick<ConditionalRulesGroupProps, 'fieldVariant'> {
  definitions: ConditionalRulesSelection[];
  name: string;
  helperText?: string;
  label?: string;
}

const useStyles = makeStyles((theme) => ({
  btn: {
    marginTop: theme.spacing(4),
  },
  fab: {
    boxShadow: 'none',
  },
  divider: {
    alignSelf: 'center',
    margin: `${theme.spacing(1)} 0`,
    height: '40px',
  },
  dividerText: {
    position: 'absolute',
    top: '19px',
    background: theme.palette.background.paper,
  },
}));

const defaultValue = [];

export const ConditionalRules = ({
  definitions,
  name,

  label,
  fieldVariant,
  helperText,
  ...rest
}: ConditionalRulesProps) => {
  const classes = useStyles();

  const [activeRole, setActiveRole] = useState<string | undefined>();
  const [activeGroupId, setActiveGroupId] = useState<string | undefined>();

  const {
    input,
    meta,
    append,
    remove,
  } = useFieldArray<ConditionalRuleGroupInput>(name, {
    ...rest,
    initialValue: rest.initialValue ?? defaultValue,
  });
  const hasError = useFieldHasError({
    meta,
    showErrorOnlyOnTouched: rest.showErrorOnlyOnTouched,
  });

  const { value } = input;

  const getFieldName = useCallback((path: string) => name + path, [name]);

  const handleAddRulesSection = useCallback(() => {
    const result = append({
      type: ConditionalRuleGroupType.Any,
      rules: [],
    });

    setActiveGroupId(result.id as string);
  }, [append]);

  return (
    <Box>
      {label && <FormLabel error={hasError}>{label}</FormLabel>}
      {helperText && <FormHelperText>{helperText}</FormHelperText>}
      <Box mt={2}>
        {value.map((group, index) => {
          const hasNext = Boolean(value[index + 1]);

          return (
            <Stack direction="column" key={group.id}>
              <ConditionalRulesGroup
                open={group.id === activeGroupId}
                onClose={() => setActiveGroupId(undefined)}
                onOpen={() => setActiveGroupId(group.id)}
                activeRowId={activeRole}
                onEdit={setActiveRole}
                onEditClose={() => setActiveRole(undefined)}
                onRemove={remove}
                name={getFieldName(`[${index}]`)}
                fieldVariant={fieldVariant}
                index={index}
                definitions={definitions}
              />
              {hasNext && (
                <Box display="flex" justifyContent="center" position="relative">
                  <Divider
                    className={classes.divider}
                    orientation="vertical"
                    flexItem
                  />
                  <Typography className={classes.dividerText} variant="caption">
                    OR
                  </Typography>
                </Box>
              )}
            </Stack>
          );
        })}
      </Box>
      <Centered className={classes.btn}>
        <Fab
          className={classNames(classes.fab, 'add-rules-group')}
          color="primary"
          size="small"
          variant="extended"
          onClick={handleAddRulesSection}
        >
          <Add />
          Add rules group
        </Fab>
      </Centered>
      {hasError && (
        <FormHelperText error={hasError}>{meta.error.message}</FormHelperText>
      )}
    </Box>
  );
};
