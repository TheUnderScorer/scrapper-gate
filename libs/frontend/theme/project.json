{
  "root": "libs/frontend/theme",
  "sourceRoot": "libs/frontend/theme/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": ["libs/frontend/theme/**/*.{ts,tsx,js,jsx}"]
      },
      "outputs": ["{options.outputFile}"]
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/frontend/theme"],
      "options": {
        "jestConfig": "libs/frontend/theme/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
