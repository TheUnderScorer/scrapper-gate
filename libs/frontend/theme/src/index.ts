export * from './ThemeProvider';
export * from './typings/emotion';
export * from './typings/material-ui';
export * from './palette';
export * from './types';
export { conditionalStopPropagation } from './conditionalStopPropagation';
export { getTheme } from './theme';
