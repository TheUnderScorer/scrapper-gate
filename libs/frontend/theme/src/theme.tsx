import { Delete, PlayArrow } from '@mui/icons-material';
import { createTheme, SimplePaletteColorOptions } from '@mui/material/styles';
import React from 'react';
import { conditionalStopPropagation } from './conditionalStopPropagation';
import { palette } from './palette';

interface ThemeParams {
  isContent?: boolean;
  container?: HTMLElement;
  htmlFontSize: number;
}

export const getTheme = ({ isContent, container, htmlFontSize }: ThemeParams) =>
  createTheme({
    palette,
    components: {
      MuiButton: {
        defaultProps: {
          disableElevation: true,
        },
        styleOverrides: {
          containedPrimary: {
            background: palette.gradients.primaryMainToDark,
            color: (palette.primary as SimplePaletteColorOptions).contrastText,
          },
        },
      },
      MuiFab: {
        styleOverrides: {
          primary: {
            '&:not[disabled]': {
              background: palette.gradients.primaryMainToDark,
            },
          },
        },
      },
      MuiPopover: {
        defaultProps: {
          disableEnforceFocus: isContent,
          container,
        },
      },
      MuiTooltip: {
        defaultProps: {
          PopperProps: {
            container,
          },
        },
      },
      MuiModal: {
        defaultProps: {
          disableEnforceFocus: isContent,
          container,
        },
      },
      MuiMenu: {
        defaultProps: {
          container,
          style: {
            pointerEvents: 'all',
          },
        },
      },
      MuiAccordion: {
        variants: [
          {
            props: {
              variant: 'transparent',
            },
            style: {
              background: 'transparent',
              border: 'none',
            },
          },
        ],
      },
      MuiDialog: {
        defaultProps: {
          container,
          disableEnforceFocus: isContent,
          disableAutoFocus: isContent,
          disableRestoreFocus: isContent,
          BackdropProps: {
            style: {
              pointerEvents: 'all',
            },
          },
          PaperProps: {
            style: {
              pointerEvents: 'all',
            },
          },
        },
      },
      MuiInput: {
        defaultProps: {
          onKeyDown: conditionalStopPropagation,
          onKeyUp: conditionalStopPropagation,
          onKeyPress: conditionalStopPropagation,
        },
      },
      MuiTextField: {
        defaultProps: {
          onKeyDown: conditionalStopPropagation,
          onKeyUp: conditionalStopPropagation,
          onKeyPress: conditionalStopPropagation,
        },
      },
    },
    typography: {
      fontSize: 13,
      htmlFontSize,
    },
    zIndex: {
      modal: 1400,
    },
    emojis: {
      empty: 'üòØ',
      error: 'üòµ',
      success: 'üòÅ',
    },
    icons: {
      run: <PlayArrow />,
      delete: <Delete />,
    },
  });
