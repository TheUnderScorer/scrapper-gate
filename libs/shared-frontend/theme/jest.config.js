module.exports = {
  displayName: 'shared-frontend-theme',
  preset: '../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../coverage/libs/shared-frontend/theme',
};
