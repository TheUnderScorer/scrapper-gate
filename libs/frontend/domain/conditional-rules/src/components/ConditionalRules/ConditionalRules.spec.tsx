import { LocalizationProvider } from '@material-ui/lab';
import AdapterDateFns from '@material-ui/lab/AdapterDateFns';
import { ThemeProvider } from '@scrapper-gate/frontend/theme';
import '@scrapper-gate/frontend/theme';
import { act, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Form } from 'react-final-form';
import { baseRulesSelection } from '../../baseRules';
import { ConditionalRules, ConditionalRulesProps } from './ConditionalRules';

const renderCmp = (props: Partial<ConditionalRulesProps> = {}) => {
  return render(
    <ThemeProvider>
      <LocalizationProvider dateAdapter={AdapterDateFns}>
        <Form
          onSubmit={jest.fn()}
          render={() => (
            <ConditionalRules
              definitions={props.definitions ?? baseRulesSelection}
              name="rules"
            />
          )}
        />
      </LocalizationProvider>
    </ThemeProvider>
  );
};

describe('<ConditionalRules />', () => {
  it('should render without crashing', () => {
    const cmp = renderCmp();

    expect(cmp).toMatchSnapshot();
  });

  it('should add new group and rule', () => {
    const cmp = renderCmp();

    act(() => {
      userEvent.click(cmp.getByText('Add rules group'));
    });

    const groups = cmp.container.querySelectorAll('.conditional-rules-group');

    expect(groups).toHaveLength(1);
    expect(groups[0].querySelector('.Mui-expanded')).toBeInTheDocument();

    act(() => {
      userEvent.click(cmp.getByText('Add rule'));
      userEvent.click(cmp.getByText('Date'));
    });

    const rules = cmp.container.querySelectorAll('.conditional-rules-rule');

    expect(rules).toHaveLength(1);
    expect(rules[0].querySelector('.Mui-expanded')).toBeInTheDocument();
  });

  it('should remove group', () => {
    const cmp = renderCmp();

    act(() => {
      userEvent.click(cmp.getByText('Add rules group'));
    });

    act(() => {
      userEvent.click(cmp.container.querySelector('.remove-rules-group'));
    });

    const groups = cmp.container.querySelectorAll('.conditional-rules-group');

    expect(groups).toHaveLength(0);
  });

  it('should remove rule', () => {
    const cmp = renderCmp();

    act(() => {
      userEvent.click(cmp.getByText('Add rules group'));
    });

    act(() => {
      userEvent.click(cmp.getByText('Add rule'));
      userEvent.click(cmp.getByText('Date'));
    });

    act(() => {
      userEvent.click(cmp.container.querySelector('.remove-rules-rule'));
    });

    const rules = cmp.container.querySelectorAll('.conditional-rules-rule');

    expect(rules).toHaveLength(0);
  });
});
