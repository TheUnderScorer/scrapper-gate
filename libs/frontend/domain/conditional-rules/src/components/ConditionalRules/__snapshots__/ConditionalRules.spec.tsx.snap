// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConditionalRules /> should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="conditional-rules MuiBox-root css-0"
        data-name="rules"
      >
        <div
          class="MuiBox-root css-0"
        />
        <div
          class="centered MuiBox-root css-7wdmh8"
        >
          <button
            class="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-sizeSmall MuiFab-secondary add-rules-group css-f121c0-MuiButtonBase-root-MuiFab-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-egrpcv-MuiSvgIcon-root"
              data-testid="AddIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
            Add rules group
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="conditional-rules MuiBox-root css-0"
      data-name="rules"
    >
      <div
        class="MuiBox-root css-0"
      />
      <div
        class="centered MuiBox-root css-7wdmh8"
      >
        <button
          class="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-sizeSmall MuiFab-secondary add-rules-group css-f121c0-MuiButtonBase-root-MuiFab-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-egrpcv-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
          Add rules group
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
